Basic audit log

soltysh opened the pull request:
Fixes #2203 by introducing simple audit logging, including the information about impersonation. We currently have something identical in openshift, but I'm open to any suggestions. Sample logs look like that:

as `<self>`:

```
AUDIT: id="75114bb5-970a-47d5-a5f1-1e99cea0574c" ip="127.0.0.1" method="GET" user="test-admin" as="<self>" namespace="openshift" uri="/api/v1/namespaces/openshift/pods/python"
AUDIT: id="75114bb5-970a-47d5-a5f1-1e99cea0574c" response=200
```

as user:

```
AUDIT: id="b0a443ae-f7d8-408c-a355-eb9501fd5c59" ip="192.168.121.118" method="GET" user="system:admin" as="test-admin" namespace="openshift" uri="/api/v1/namespaces/openshift/pods/python"
AUDIT: id="b0a443ae-f7d8-408c-a355-eb9501fd5c59" response=200
```

``` release-note
* Add basic audit logging
```

@ericchiang @smarterclayton @roberthbailey @erictune @ghodss 

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27087)

<!-- Reviewable:end -->

soltysh committed:
Basic audit log

ericchiang commented on pkg/genericapiserver/genericapiserver.go:
```diff
@@ -475,6 +479,10 @@ func (s *GenericAPIServer) init(c *Config) {
 	handler = apiserver.WithAuthorizationCheck(handler, attributeGetter, s.authorizer)
+	if c.EnableAudit {
+		// audit handler must comes before the impersonationFilter to read the original user
+		handler = audit.WithAudit(handler, s.RequestContextMapper)
```
Should this come before authorization? I imagine we'd want to capture authentication requests that were denied for some reason or another.

ericchiang replied:
Woops, sorry it already does. (Got my middleware logic backwards)

ericchiang commented:
This is great. Few comments:

* Would like to see a pluggable interface for this. We'd like implementations other than stderr logging pretty quickly.

* I don't see how we could extend this to provide information such as specifics around authenticator, authorizer, or admission controller. For example, which authentication strategy was used, or which admission controller denied a particular request. This was brought up in the initial feature specs[0], but I don't know if that's too much trouble to try to accomplish.

[0] https://github.com/kubernetes/kubernetes/issues/2203#issue-47993283

lavalamp commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func (a *auditResponseWriter) WriteHeader(code int) {
+	glog.Infof("AUDIT: id=%q response=\"%d\"", a.id, code)
```
Is writing to stdout really the best way to produce audit logs? I guess I'd expect a separate output file, though I've never made audit logs before.

smarterclayton replied:
A separate log file is not uncommon (depends on whether you want your logs
in one place or two places) - putting this behind an interface certainly
can't hurt though, and fleshing out the implementation.

On Wed, Jun 8, 2016 at 6:56 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/apiserver/audit/audit.go
> <https://github.com/kubernetes/kubernetes/pull/27087#discussion_r66355213>
> :
>
> > +
> > +	"github.com/golang/glog"
> > +	"github.com/pborman/uuid"
> > +
> > +	"k8s.io/kubernetes/pkg/api"
> > +	utilnet "k8s.io/kubernetes/pkg/util/net"
> > +)
> > +
> > +// auditResponseWriter implements http.ResponseWriter interface.
> > +type auditResponseWriter struct {
> > +	http.ResponseWriter
> > +	id string
> > +}
> > +
> > +func (a *auditResponseWriter) WriteHeader(code int) {
> > +	glog.Infof("AUDIT: id=%q response=\"%d\"", a.id, code)
>
> Is writing to stdout really the best way to produce audit logs? I guess
> I'd expect a separate output file, though I've never made audit logs before.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27087/files/752c551c39be25f16e03d37bec8a3c2567467370#r66355213>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p4gePInwAQlDTfd0K-tGyId2SyZeks5qJ0iqgaJpZM4Ixcs->
> .
>

lavalamp replied:
The log is already hard enough to read-- I think I'd prefer a separate file.

Also I think it might be good to put the audit ID in the context? I guess
we can do that when we have a use case.

On Wed, Jun 8, 2016 at 4:07 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/apiserver/audit/audit.go
> <https://github.com/kubernetes/kubernetes/pull/27087#discussion_r66356421>
> :
>
> > +
> > +	"github.com/golang/glog"
> > +	"github.com/pborman/uuid"
> > +
> > +	"k8s.io/kubernetes/pkg/api"
> > +	utilnet "k8s.io/kubernetes/pkg/util/net"
> > +)
> > +
> > +// auditResponseWriter implements http.ResponseWriter interface.
> > +type auditResponseWriter struct {
> > +	http.ResponseWriter
> > +	id string
> > +}
> > +
> > +func (a *auditResponseWriter) WriteHeader(code int) {
> > +	glog.Infof("AUDIT: id=%q response=\"%d\"", a.id, code)
>
> A separate log file is not uncommon (depends on whether you want your logs
> in one place or two places) - putting this behind an interface certainly
> can't hurt though, and fleshing out the implementation.
> … <#m_-3854526224780300919_>
> On Wed, Jun 8, 2016 at 6:56 PM, Daniel Smith ***@***.***> wrote: In
> pkg/apiserver/audit/audit.go <#27087 (comment)
> <https://github.com/kubernetes/kubernetes/pull/27087#discussion_r66355213>>
> : > + > + "github.com/golang/glog" > + "github.com/pborman/uuid" > + > + "
> k8s.io/kubernetes/pkg/api" > + utilnet "k8s.io/kubernetes/pkg/util/net" >
> +) > + > +// auditResponseWriter implements http.ResponseWriter interface.
> > +type auditResponseWriter struct { > + http.ResponseWriter > + id string
> > +} > + > +func (a *auditResponseWriter) WriteHeader(code int) { > +
> glog.Infof("AUDIT: id=%q response=\"%d\"", a.id, code) Is writing to
> stdout really the best way to produce audit logs? I guess I'd expect a
> separate output file, though I've never made audit logs before. — You are
> receiving this because you were mentioned. Reply to this email directly,
> view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/27087/files/752c551c39be25f16e03d37bec8a3c2567467370#r66355213>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/ABG_p4gePInwAQlDTfd0K-tGyId2SyZeks5qJ0iqgaJpZM4Ixcs->
> .
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27087/files/752c551c39be25f16e03d37bec8a3c2567467370#r66356421>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAngljnZIzH5MvZkcNtqBs9oe67OxgN5ks5qJ0sjgaJpZM4Ixcs->
> .
>

soltysh replied:
> The log is already hard enough to read-- I think I'd prefer a separate file. 

I don't mind that, can you please tell me how to do it with glog though? I haven't see any possibility there :( But I agree with Clayton about placing it behind an interface. 

> Also I think it might be good to put the audit ID in the context? I guess we can do that when we have a use case.

Makes sense as well.

lavalamp replied:
Just don't use glog may be the easiest :)

On Thu, Jun 9, 2016 at 7:36 AM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/apiserver/audit/audit.go
> <https://github.com/kubernetes/kubernetes/pull/27087#discussion_r66452677>
> :
>
> > +
> > +	"github.com/golang/glog"
> > +	"github.com/pborman/uuid"
> > +
> > +	"k8s.io/kubernetes/pkg/api"
> > +	utilnet "k8s.io/kubernetes/pkg/util/net"
> > +)
> > +
> > +// auditResponseWriter implements http.ResponseWriter interface.
> > +type auditResponseWriter struct {
> > +	http.ResponseWriter
> > +	id string
> > +}
> > +
> > +func (a *auditResponseWriter) WriteHeader(code int) {
> > +	glog.Infof("AUDIT: id=%q response=\"%d\"", a.id, code)
>
> The log is already hard enough to read-- I think I'd prefer a separate
> file.
>
> I don't mind that, can you please tell me how to do it with glog though? I
> haven't see any possibility there :( But I agree with Clayton about placing
> it behind an interface.
>
> Also I think it might be good to put the audit ID in the context? I guess
> we can do that when we have a use case.
>
> Makes sense as well.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27087/files/752c551c39be25f16e03d37bec8a3c2567467370#r66452677>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglhR5eD-2XOT8-YcGg8_0JQ1dW5Myks5qKCTVgaJpZM4Ixcs->
> .
>

soltysh replied:
I'll go with Clayton's solution of passing a writer, how that will be resolved further we can leave for other time :)

erictune replied:
Separate file seems like a good default.  

I'm going to want to add an interface that can call out to a webhook to post a batch of these.  But that can wait for a later PR.  Going with passing Writer in this PR seems fine.

lavalamp commented on pkg/genericapiserver/genericapiserver.go:
```diff
@@ -197,6 +199,7 @@ type GenericAPIServer struct {
 	MuxHelper             *apiserver.MuxHelper
 	HandlerContainer      *restful.Container
 	RootWebService        *restful.WebService
+	enableAudit           bool
```
I don't think you ever reference this? remove?

lavalamp commented:
So I think the most important thing about audit logging is probably certainty that it always gets called for all api accesses. I think you need a testing strategy for this.

I also think this probably gives incorrect or unhelpful output for watches, port forwarding / attach / exec, proxy requests.

erictune commented:
I like this.  Thanks for writing it.

roberthbailey commented:
/cc @cjcullen

erictune commented:
@maisem you have some expertise in this area: would you take a look at the list of things that are being audit logged?

erictune commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,109 @@
+//    - impersonated user for the operation
+//    - namespace of the request or <none>
+//    - uri is the full URI as requested
+// 2. the response line containing the unique id from 1 and response code
```
@maisem take a look at above comment

maisem replied:
Can you split this line into the same format as 1. 
As it is right now, I thought it meant that we log the response body.

maisem replied:
I guess my question is should we log the response/request body as well?

soltysh replied:
> Can you split this line into the same format as 1. 
> As it is right now, I thought it meant that we log the response body.

Not sure what you need here. Basically it's split into two lines because of the response code. I want to have two separate logs for the matter of knowing which and when went in. I don't see any value in repeating all data in the 2nd line. 

> I guess my question is should we log the response/request body as well?

If you want to have full request/response body just increase the loglevel. This is meant rather as just a general audit information, hopefully at some point in time landing in a separate file.

maisem replied:
Sorry, I meant the comment line.

```
2. the response line containing:
  - the unique id from 1 
  - response code
```

soltysh replied:
:blush: sure thing

erictune replied:
Logging the whole body for a secret create request seems like a bad idea.

On Thu, Jun 30, 2016 at 7:34 AM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/apiserver/audit/audit.go
> <https://github.com/kubernetes/kubernetes/pull/27087#discussion_r69143282>
> :
>
> > +
> > +var _ http.CloseNotifier = &fancyResponseWriterDelegator{}
> > +var _ http.Flusher = &fancyResponseWriterDelegator{}
> > +var _ http.Hijacker = &fancyResponseWriterDelegator{}
> > +
> > +// WithAudit is responsible for logging audit information for all the
> > +// request coming to server. Each audit log contains two entries:
> > +// 1. the request line containing:
> > +//    - unique id allowing to match the response line (see 2)
> > +//    - source ip of the request
> > +//    - HTTP method being invoked
> > +//    - original user invoking the operation
> > +//    - impersonated user for the operation
> > +//    - namespace of the request or <none>
> > +//    - uri is the full URI as requested
> > +// 2. the response line containing the unique id from 1 and response code
>
> 😊 sure thing
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27087/files/752c551c39be25f16e03d37bec8a3c2567467370#r69143282>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHuudsn2QQ9HN0VVhLSrMxG9Krok0TzZks5qQ9P5gaJpZM4Ixcs->
> .
>

maisem replied:
I agree, but that's solvable by redacting some of the fields.
It would be good to know what the caller was generating secrets for.

It could be nice if the id generated here was available for log correlation at the api level.

soltysh replied:
Yet again, that falls into the category of debugging server, which can be easily achieved with loglevel=8. Audit is different, imho.

maisem commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,109 @@
+		user, _ := api.UserFrom(ctx)
+		asuser := req.Header.Get("Impersonate-User")
+		if len(asuser) == 0 {
+			asuser = "<self>"
```
should this be `user.GetName()` instead of `<self>`.

soltysh replied:
imho, `<self>` is easier to grep than having double value for the username.

sttts replied:
I think the usual use-case for grepping is by `as=xyz`, not by `user=xyz as=<self>`. I prefer less magic and double value in the log.

lavalamp replied:
That seems like a solid argument. +1

On Thu, Jul 14, 2016 at 1:27 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In pkg/apiserver/audit/audit.go
> <https://github.com/kubernetes/kubernetes/pull/27087#discussion_r70767483>
> :
>
> > +// 1. the request line containing:
> > +//    - unique id allowing to match the response line (see 2)
> > +//    - source ip of the request
> > +//    - HTTP method being invoked
> > +//    - original user invoking the operation
> > +//    - impersonated user for the operation
> > +//    - namespace of the request or <none>
> > +//    - uri is the full URI as requested
> > +// 2. the response line containing the unique id from 1 and response code
> > +func WithAudit(handler http.Handler, requestContextMapper api.RequestContextMapper) http.Handler {
> > +	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
> > +		ctx, _ := requestContextMapper.Get(req)
> > +		user, _ := api.UserFrom(ctx)
> > +		asuser := req.Header.Get("Impersonate-User")
> > +		if len(asuser) == 0 {
> > +			asuser = "<self>"
>
> I think the usual use-case for grepping is by as=xyz, not by user=xyz
> as=<self>. I prefer less magic and double value in the log.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27087/files/752c551c39be25f16e03d37bec8a3c2567467370#r70767483>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglsLx0miG4g8PL__wAlulXsyYN9Xgks5qVfMDgaJpZM4Ixcs->
> .
>

maisem commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,109 @@
+//    - original user invoking the operation
+//    - impersonated user for the operation
+//    - namespace of the request or <none>
+//    - uri is the full URI as requested
```
Should we also log the request body?

soltysh commented:
I'll try to update this PR tomorrow, now that 1.4 queue is open.

soltysh commented:
Updated, ptal.

amouat commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+func (a *auditResponseWriter) WriteHeader(code int) {
+	fmt.Fprintf(a.out, "AUDIT: id=%q response=\"%d\"", a.id, code)
```
Would it make sense to add a newline here? At the moment all the logs appear in a single line which is a bit of a pain.

soltysh replied:
True, I totally forgot about that one when switching from glog to Fprintf.

lavalamp replied:
+1 for newline

soltysh replied:
Done.

mwl commented on pkg/genericapiserver/genericapiserver.go:
```diff
@@ -475,6 +477,14 @@ func (s *GenericAPIServer) init(c *Config) {
 	attributeGetter := apiserver.NewRequestAttributeGetter(s.RequestContextMapper, s.NewRequestInfoResolver())
 	handler = apiserver.WithAuthorizationCheck(handler, attributeGetter, s.authorizer)
+	if len(c.AuditLog) != 0 {
+		auditLog, err := os.OpenFile(c.AuditLog, os.O_APPEND|os.O_WRONLY, 0644)
```
I believe `O_CREATE` also should be allowed, in cases where the file doesn't exists already? When running with Hyperkube I get the following error
> F0707 12:42:48.410710       1 genericapiserver.go:483] Could not open audit log file: open /kube-audit.log: no such file or directory

mwl replied:
@soltysh please see soltysh/kubernetes#1

soltysh replied:
Done.

alexbrand commented:
Should the audit log messages include a timestamp?

lavalamp commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,114 @@
+		}
+		id := uuid.NewRandom().String()
+
+		fmt.Fprintf(out, "AUDIT: id=%q ip=%q method=%q user=%q as=%q namespace=%q uri=%q",
```
What do you think about making this more structured, e.g. json or yaml?

soltysh replied:
What do you mean by that? Printing yaml/json here, eg:
```yaml
audit:
  id: 123.456
  ip: 192.168.1.1
  method: POST
...
```
Wouldn't that make the log too long? With above you have just one line for each call, easily greppable.

ghodss replied:
I think key-value as it's written is a good first option since it's most amenable to systems like splunk and other conventional security logging utilities. If any other formats are added (e.g. JSON blobs on one one line) they should be a configuration option.

lavalamp replied:
Ah, ok. If this is a common format then that's totally fine with me.

On Tue, Jul 12, 2016 at 5:01 PM, Sam Ghods <notifications@github.com> wrote:

> In pkg/apiserver/audit/audit.go
> <https://github.com/kubernetes/kubernetes/pull/27087#discussion_r70544596>
> :
>
> > +//    - response code
> > +func WithAudit(handler http.Handler, requestContextMapper api.RequestContextMapper, out io.Writer) http.Handler {
> > +	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
> > +		ctx, _ := requestContextMapper.Get(req)
> > +		user, _ := api.UserFrom(ctx)
> > +		asuser := req.Header.Get("Impersonate-User")
> > +		if len(asuser) == 0 {
> > +			asuser = "<self>"
> > +		}
> > +		namespace := api.NamespaceValue(ctx)
> > +		if len(namespace) == 0 {
> > +			namespace = "<none>"
> > +		}
> > +		id := uuid.NewRandom().String()
> > +
> > +		fmt.Fprintf(out, "AUDIT: id=%q ip=%q method=%q user=%q as=%q namespace=%q uri=%q",
>
> I think key-value as it's written is a good first option since it's most
> amenable to systems like splunk and other conventional security logging
> utilities. If any other formats are added (e.g. JSON blobs on one one line)
> they should be a configuration option.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27087/files/39a81bd9c13e88bfa0da8c2bc6f91cf4ae450a28#r70544596>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglp0vFSO0I3GTyjzzU862gyP4AwbJks5qVCrpgaJpZM4Ixcs->
> .
>

lavalamp commented:
+1 for timestapm. Looks close. It would be super awesome if you made the ordinary log messages include the audit log ID-- that would make it easier to trace a crash back to a request.

I really want this because right now we've got no way to see request sources for requests that are still ongoing.

soltysh commented:
@lavalamp yeah, I'll try to address the suggestions tomorrow. 

> It would be super awesome if you made the ordinary log messages include the audit log ID-- that would make it easier to trace a crash back to a request.

Can you elaborate a bit more what do you want here? Not sure I follow.

amouat commented:
@soltysh @lavalamp I think I understand, but I would suggest it is out of scope for this. The ID here is necessary to correlate the request and the response. @lavalamp is suggesting that all k8s logs that can be traced back to a request include this ID. If this was done, an error in any part of k8s could be traced back to the API call that triggered it. I'm not sure this is feasible, but it's certainly a nice idea.

ncdc commented:
FYI @sttts - we need to work with @soltysh on this to include the scope of your card, or do it as a follow-up

lavalamp commented:
Mostly I was just thinking that having the ID with the stack traces that it prints out if it crashes or returns an unexpected status code would be handy. Doesn't need to be in everything.

As for yaml, json can go all on one line if you're worried about length. But is it really a problem to have line breaks in a log message? I'm just thinking from the perspective of someone who wants to datamine these logs.

soltysh commented:
I've added timestamps, newlines and fixed the file access mode. That's that much time I can commit working on this PR, at least now. I'd propose merging this as is and following-up with other improvements, does that make sense?

lavalamp commented:
Now that I think about this, I want to know the strategy for rotating/archiving this audit file? It's going to get huge, fast...

Should it really be a separate file? Kubernetes's own rotation system only works on stdout/err IIRC.

erictune commented:
The natural next step after this is to have an audit logging interface and
be able to log the same set of data to a variety of different file formats
and/or network-sinks.

On Wed, Jul 13, 2016 at 9:12 AM, Daniel Smith <notifications@github.com>
wrote:

> Now that I think about this, I want to know the strategy for
> rotating/archiving this audit file? It's going to get huge, fast...
>
> Should it really be a separate file? Kubernetes's own rotation system only
> works on stdout/err IIRC.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27087#issuecomment-232405350>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHuudnHtou11mHQP-Er5KcFYT1HrSXG4ks5qVQ5ugaJpZM4Ixcs->
> .
>

soltysh commented:
This boils down the the limitation of glog, which does not allow having multiple log files (eg. one per each component), which would be desired here, imho. I'm afraid we're going in direction of inventing another logging library :/

sttts commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,115 @@
+	utilnet "k8s.io/kubernetes/pkg/util/net"
+)
+
+// auditResponseWriter implements http.ResponseWriter interface.
```
just move the `var _ ...` up here and we can get rid of the comment ;-)

sttts commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,115 @@
+var _ http.Flusher = &fancyResponseWriterDelegator{}
+var _ http.Hijacker = &fancyResponseWriterDelegator{}
+
+// WithAudit is responsible for logging audit information for all the
```
nit: WithAudit decorates a http.Handler with audit logging....

sttts commented on pkg/apiserver/audit/audit.go:
```diff
@@ -0,0 +1,115 @@
+	})
+}
+
+func constructResponseWriter(responseWriter http.ResponseWriter, out io.Writer, id string) http.ResponseWriter {
```
Isn't this what you usually call `NewResponseWriter` ?

sttts replied:
or maybe `decorateResponseWriter`. `construct` is pretty generic.

sttts commented on pkg/genericapiserver/genericapiserver.go:
```diff
@@ -94,6 +95,7 @@ type APIGroupInfo struct {
 type Config struct {
 	// The storage factory for other objects
 	StorageFactory StorageFactory
+	AuditLog       string
```
nit: AuditLogPath

sttts commented on pkg/genericapiserver/genericapiserver.go:
```diff
@@ -475,6 +477,14 @@ func (s *GenericAPIServer) init(c *Config) {
 	attributeGetter := apiserver.NewRequestAttributeGetter(s.RequestContextMapper, s.NewRequestInfoResolver())
 	handler = apiserver.WithAuthorizationCheck(handler, attributeGetter, s.authorizer)
+	if len(c.AuditLog) != 0 {
+		auditLog, err := os.OpenFile(c.AuditLog, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
```
In contrib/mesos (https://github.com/kubernetes/kubernetes/blob/master/contrib/mesos/pkg/minion/server.go#L226) we use lumberjack.v2 for log rotation. Maybe that's an option here as well.

erictune commented:
I don't see it as inventing a logging library, since the set of data is fixed.  glog is not going to work for sending audit logs to a cloud audit logging sink.

sttts commented:
@soltysh From a discussion in yesterday's sig-api-machinery meeting: what is mainly open here is only log rotation.

soltysh commented:
@sttts I'll be back tomorrow and I'll try to address this either tomorrow or at latest next week to be able to close this before my PTO in mid Aug, which also is before the 1.4 code freeze. Thanks for the info!

soltysh commented:
Rebased and comments addressed @sttts mind taking a look?

sttts commented:
Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/apiserver/audit/audit.go, line 85 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtK6kMxwZW4wMISjAN:-KOtK6kMxwZW4wMISjAO:b1n9gmf) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/apiserver/audit/audit.go#L85)):*
> ```Go
> 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
> 		ctx, _ := requestContextMapper.Get(req)
> 		user, _ := api.UserFrom(ctx)
> ```

Can user be empty? We get the empty string then in the logs?

---

*[pkg/apiserver/audit/audit_test.go, line 47 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtLO9t1Dpf5JiYQH7E:-KOtLO9t1Dpf5JiYQH7F:b-mapl6c) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/apiserver/audit/audit_test.go#L47)):*
> ```Go
> 	switch v := actual.(type) {
> 	case *auditResponseWriter:
> 		break
> ```

nit: no break necessary in go

---

*[pkg/genericapiserver/genericapiserver.go, line 483 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtJQlXKnlD4-cyur8u:-KOtJQlXKnlD4-cyur8v:b-g7w5xr) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/genericapiserver/genericapiserver.go#L483)):*
> ```Go
> 	if len(c.AuditLogPath) != 0 {
> 		// audit handler must comes before the impersonationFilter to read the original user
> 		writer := &lumberjack.Logger{Filename: c.AuditLogPath}
> ```

We should add --audit-log-max-size/max-backup/.... But that can wait for another PR. The default is 100MB, then old logs are removed.

---

*[pkg/genericapiserver/options/server_run_options.go, line 299 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtJrb-WbwJKn9IeMzB:-KOtJrb-WbwJKn9IeMzC:b-25y7m0) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/genericapiserver/options/server_run_options.go#L299)):*
> ```Go
> 
> 	fs.StringVar(&s.AuditLogPath, "audit-log", s.AuditLogPath,
> 		"If set, all requests coming to the apiserver will be logged to this file.")
> ```

Maybe a word about log rotation?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

sttts commented:
One nit, and a question about empty `user`. Otherwise, looks good to me.

---

Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-:-KOtMHr1nwTYrwZdgMbx:b-f9o6en)*
<!-- Sent from Reviewable.io -->

soltysh commented:
*[pkg/apiserver/audit/audit.go, line 85 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtK6kMxwZW4wMISjAN:-KOtl940XYFC1FSzqGhS:br3nlz6) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/apiserver/audit/audit.go#L85)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary>
> Can user be empty? We get the empty string then in the logs?
</details>
It should not, because we're doing this only against the secured entrypoint. The insecured one is not monitored. If we want to we should, I'd propose opening a follow-up.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

soltysh commented:
*[pkg/apiserver/audit/audit_test.go, line 47 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtLO9t1Dpf5JiYQH7E:-KOtldhkM91TQ8uKhXhz:b5zf1gt) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/apiserver/audit/audit_test.go#L47)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary>
> nit: no break necessary in go
</details>
This one is needed, since we want to exclude this particular type, without it, it'll fall under default action, which is not what we want.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

soltysh commented:
*[pkg/genericapiserver/genericapiserver.go, line 483 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtJQlXKnlD4-cyur8u:-KOtlsmfgh1xyVg3MT9d:b-pbdqz5) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/genericapiserver/genericapiserver.go#L483)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary>
> We should add --audit-log-max-size/max-backup/.... But that can wait for another PR. The default is 100MB, then old logs are removed.
</details>
Agreed,

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

soltysh commented:
@sttts added info about rotation and commented about user, although it's terrible with reviewable :disappointed:

sttts commented:
Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/genericapiserver/options/server_run_options.go, line 299 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtJrb-WbwJKn9IeMzB:-KOton-RE3YEz3HBOMRJ:b-a8da5x) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/genericapiserver/options/server_run_options.go#L299)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary>
> Maybe a word about log rotation?
</details>
Default for MaxAge and MaxBackup is unlimited: https://github.com/natefinch/lumberjack#type-logger. We should set sane values, e.g. 7 days.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

deads2k commented:
wow, reviewable makes some ugly comments in github

sttts commented:
Oh yes, it does.

---

Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-:-KOtpwrfOUAZdeVcYomW:b-b0a6xz)*
<!-- Sent from Reviewable.io -->

deads2k commented:
*[pkg/apiserver/audit/audit_test.go, line 47 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtLO9t1Dpf5JiYQH7E:-KOtroJPVHoet_IsK9Ph:b9jq2tb) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/apiserver/audit/audit_test.go#L47)):*
<details><summary><i>Previously, soltysh (Maciej Szulik) wrote…</i></summary>
> This one is needed, since we want to exclude this particular type, without it, it'll fall under default action, which is not what we want.
</details>
Go breaks at the end of the block by default.  You have to specify `fallthrough` to drop through.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

soltysh commented:
@sttts added the 3 options (maxbackup, maxage and maxsize) to configure the logging as discussed on IRC and fixed the break, as well. ptal

sttts commented:
lgtm

---

Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-:-KOtvvoJzfnIcNEwrZwT:bb7k0qh)*
<!-- Sent from Reviewable.io -->

soltysh commented:
Fixed boilerplate in the newly added files, which was causing jenkins failure. Applying label based on previous approval.

soltysh commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/apiserver/audit/audit_test.go, line 47 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtLO9t1Dpf5JiYQH7E:-KOxuG3koEQbNQ4VBKtN:b-886gsx) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/apiserver/audit/audit_test.go#L47)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary>
> Go breaks at the end of the block by default.  You have to specify `fallthrough` to drop through.
</details>
Fixed

---

*[pkg/genericapiserver/options/server_run_options.go, line 299 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtJrb-WbwJKn9IeMzB:-KOxuKbbO4fScksRv1rf:bx3bjxz) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/genericapiserver/options/server_run_options.go#L299)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary>
> Default for MaxAge and MaxBackup is unlimited: https://github.com/natefinch/lumberjack#type-logger. We should set sane values, e.g. 7 days.
</details>
Added `--audit-log-maxage`, `--audit-log-maxbackup` and `--audit-log-maxsize` to address that.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

soltysh commented:
*[pkg/genericapiserver/options/server_run_options.go, line 299 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087#-KOtJrb-WbwJKn9IeMzB:-KOxvAPsiJGNU3WvTaYP:b32w0x) ([raw file](https://github.com/kubernetes/kubernetes/blob/1a652fdc952f100b196a2420e16eab8e2c8c2400/pkg/genericapiserver/options/server_run_options.go#L299)):*
<details><summary><i>Previously, soltysh (Maciej Szulik) wrote…</i></summary>
> Added `--audit-log-maxage`, `--audit-log-maxbackup` and `--audit-log-maxsize` to address that.
</details>
OK

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

soltysh commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->

soltysh commented:
Rebased and fixed the golint error.

sttts commented:
Reviewed 4 of 4 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/27087)*
<!-- Sent from Reviewable.io -->
